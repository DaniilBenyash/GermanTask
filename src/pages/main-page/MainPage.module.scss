.main_page {
  min-height: 100vh;
  background-color: #ebecec;
}
.section {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
  gap: 39px;
  padding: 48px;
}
.cards_section {
  display: grid;
  gap: 20px;
  grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
}

.mobile_button {
  display: none;
}
.box_contact {
  width: 100%;
}
@media screen and (max-width: 500px) {
  .section {
    gap: 0;
    padding: 0;
  }

  .cards_section {
    width: 100%;
    gap: 0;
    grid-template-columns: 1fr;
  }
  .desktop_button {
    display: none;
  }
  .mobile_button {
    display: block;
    position: fixed;
    bottom: 30px;
    right: 30px;
  }
}
