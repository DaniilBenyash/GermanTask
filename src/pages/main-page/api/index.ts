import { createAPI } from "../../helpers/createApi";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

export interface Contact {
  id: number;
  firstname: string;
  lastname: string;
  email: string;
}

const contactsKeys = {
  contacts: ["contacts"],
};

const contactsApi = createAPI("/contacts");

export const useGetUsers = () => {
  return useQuery({
    queryKey: contactsKeys.contacts,
    queryFn: async () => {
      const response = await contactsApi.get("/");
      return response.data.responseObject as Contact[];
    },
  });
};

export const useAddUser = () => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (contact: Contact) => contactsApi.post("/", contact),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: contactsKeys.contacts });
    },
  });
};

export const useUpdateUser = () => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (updatedUser: Contact) => contactsApi.put(`/${updatedUser.id}`, updatedUser),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: contactsKeys.contacts });
    },
  });
};

export const useDeleteUser = () => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: number) => contactsApi.delete(`/${id}`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: contactsKeys.contacts });
    },
  });
};
