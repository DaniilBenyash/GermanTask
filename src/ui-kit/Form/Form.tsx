import { FC, ReactNode } from "react";
import styles from "./Form.module.scss";

type FormProps = {
  onSubmit: () => void;
  children: ReactNode;
};
export const Form: FC<FormProps> = ({ onSubmit, children }) => {
  return (
    <form className={styles.form} onSubmit={onSubmit}>
      {children}
    </form>
  );
};
