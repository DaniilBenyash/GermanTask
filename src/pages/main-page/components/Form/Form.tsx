import { FC } from "react";
import styles from "./Form.module.scss";
import { Contact } from "../../api";
import { SubmitHandler, useForm } from "react-hook-form";
import { Button, Input, Form as FormUI } from "../../../../ui-kit";

type FormProps = {
  contact: Partial<Contact> | null;
  onAddUser: (contact: Contact) => void;
  onUpdateUser: (contact: Contact) => void;
  onDeleteUser: (id: number) => void;
};

export const Form: FC<FormProps> = ({ contact, onAddUser, onDeleteUser, onUpdateUser }) => {
  const {
    register,
    handleSubmit,
    reset,
    formState: { isValid },
  } = useForm<Contact>({
    defaultValues: { ...contact },
  });
  const onSubmit: SubmitHandler<Contact> = (contact) => {
    if (!contact.id) {
      onAddUser(contact);
      return;
    }
    onUpdateUser(contact);
  };

  const handleDelete = () => {
    if (!contact?.id) return;
    onDeleteUser(contact.id);
  };
  return (
    <FormUI onSubmit={handleSubmit(onSubmit)}>
      <div className={styles.data_container}>
        <h2 className={styles.title}>Eintrag bearbeiten</h2>
        <div className={styles.input_container}>
          <Input placeholder="Vorname*" {...register("firstname", { required: true })} />
          <Input placeholder="Nachname*" {...register("lastname", { required: true })} />
          <Input placeholder="E-Mail*" {...register("email", { required: true })} />
        </div>
      </div>
      <div className={styles.button_container}>
        {contact?.id && (
          <Button type="secondary" onClick={handleDelete}>
            LÃ–SCHEN
          </Button>
        )}
        <div className={styles.right_section_buttons}>
          <Button type="primary">ABBRECHEN</Button>
          <Button type="default" disabled={!isValid}>
            SPEICHERN
          </Button>
        </div>
      </div>
    </FormUI>
  );
};
