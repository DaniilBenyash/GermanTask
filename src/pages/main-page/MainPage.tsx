import styles from "./MainPage.module.scss";
import { Header, Card, Modal, Button } from "../../components";
import { Form } from "./components/Form";
import { useState } from "react";
import { useAddUser, useDeleteUser, useGetUsers, useUpdateUser } from "./api";
import { Contact } from "../../types/types";

export const MainPage = () => {
  const { data } = useGetUsers();
  const { mutate: onAddUser } = useAddUser();
  const { mutate: onUpdateUser } = useUpdateUser();
  const { mutate: onDeleteUser } = useDeleteUser();
  const [isOpenModal, setIsOpenModal] = useState(false);
  const [contact, setContact] = useState<Contact | null>(null);

  const handleCloseModal = () => {
    setContact(null);
    setIsOpenModal(false);
  };
  const handleOpenModal = () => {
    setIsOpenModal(true);
  };
  const handleOpenModalWithData = (contact: Contact) => {
    setContact(contact);
    setIsOpenModal(true);
  };

  const isMobile = window.innerWidth < 534;

  return (
    <main className={styles.main_page}>
      <Header />
      <section className={styles.section}>
        <Button className={isMobile ? styles.mobile_button : styles.desktop_button} onClick={handleOpenModal}>
          {isMobile ? <p className={styles.plus}>+</p> : 'NEUER EINTRAG'}
        </Button>
        <div className={styles.cards_section}>
          {data?.map((contact) => {
            return (
              <Card
                key={contact.id}
                title={`${contact.firstname} ${contact.lastname}`}
                email={contact.email}
                onClick={() => handleOpenModalWithData(contact)}
              />
            );
          })}
        </div>
      </section>
      <Modal isOpen={isOpenModal} onClose={handleCloseModal}>
        <Form
          contact={contact}
          onUpdateUser={onUpdateUser}
          onAddUser={onAddUser}
          onDeleteUser={onDeleteUser}
          onCloseModal={handleCloseModal}
        />
      </Modal>
    </main>
  );
};
