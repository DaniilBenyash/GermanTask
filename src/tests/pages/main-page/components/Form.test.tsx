import { Contact } from "../../../../pages/main-page/api";
import { Form } from "../../../../pages/main-page/components/Form";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Form
        contact={null}
        onAddUser={function (contact: Contact): void {
          throw new Error("Function not implemented.");
        }}
        onUpdateUser={function (contact: Contact): void {
          throw new Error("Function not implemented.");
        }}
        onDeleteUser={function (id: number): void {
          throw new Error("Function not implemented.");
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
