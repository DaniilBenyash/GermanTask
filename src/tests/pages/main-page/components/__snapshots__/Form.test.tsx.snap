// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="data_container"
  >
    <h2
      className="title"
    >
      Eintrag bearbeiten
    </h2>
    <div
      className="input_container"
    >
      <div
        className="container"
      >
        <input
          className="input"
          name="firstname"
          onChange={[Function]}
          placeholder=""
        />
        <label
          className="label"
        >
          Vorname*
        </label>
      </div>
      <div
        className="container"
      >
        <input
          className="input"
          name="lastname"
          onChange={[Function]}
          placeholder=""
        />
        <label
          className="label"
        >
          Nachname*
        </label>
      </div>
      <div
        className="container"
      >
        <input
          className="input"
          name="email"
          onChange={[Function]}
          placeholder=""
        />
        <label
          className="label"
        >
          E-Mail*
        </label>
      </div>
    </div>
  </div>
  <div
    className="button_container"
  >
    <div
      className="right_section_buttons"
    >
      <button
        className="button button_primary"
      >
        ABBRECHEN
      </button>
      <button
        className="button button_default"
        disabled={true}
      >
        SPEICHERN
      </button>
    </div>
  </div>
</form>
`;
